src/durbin_watson.jl

```diff
@@ -0,0 +1,242 @@
+  [http://www.jstor.org/stable/2332325](http://www.jstor.org/stable/2332325).
+  * J. Durbin and G. S. Watson, 1950, "Testing for Serial Correlation in Least Squares
+  Regression: I", Biometrika, Vol. 37, No. 3/4, pp. 409-428,
+  [http://www.jstor.org/stable/2332391](http://www.jstor.org/stable/2332391).
+  * R. W. Farebrother, 1980, "Algorithm AS 153: Pan's Procedure for the Tail Probabilities
+  of the Durbin-Watson Statistic", Journal of the Royal Statistical Society, Series C
+  (Applied Statistics), Vol. 29, No. 2, pp. 224-227,
+  [http://www.jstor.org/stable/2986316](http://www.jstor.org/stable/2986316).
+
+# External links
+
+  * [Durbin-Watson test on Wikipedia:
+  https://en.wikipedia.org/wiki/Durbin–Watson_statistic
+  ](https://en.wikipedia.org/wiki/Durbin–Watson_statistic)
+"""
+function DurbinWatsonTest{T<:Real}(xmat::AbstractArray{T}, e::AbstractArray{T};
+    p_compute::Symbol = :ndep)
+
+    n = length(e)
+    DW = sum(diff(e) .^2) / sum(e .^2)
+
+    DurbinWatsonTest(xmat, n, DW, p_compute)
+end
+
+testname(::DurbinWatsonTest) = "Durbin-Watson autocorrelation test"
+population_param_of_interest(x::DurbinWatsonTest) =
+    ("sample autocorrelation parameter", "0", 1 - x.DW / 2)
+default_tail(test::DurbinWatsonTest) = :both
+
+function show_params(io::IO, x::DurbinWatsonTest, ident)
+    println(io, ident, "number of observations:     ", x.n)
+    println(io, ident, "DW statistic:               ", x.DW)
+end
+
+"""
+    pan_algorithm(a::AbstractArray, x::Float64, m::Int, n::Int)
+
+Compute exact p-values for the Durbin-Watson statistic using Pan's algorithm (Farebrother,
+1980).
+
+`a` is the vector of non-zero Eigenvalues of ``(I-(X(X'X)^{-1}X'))A`` (see Durbin and
+Watson, 1971, p. 2), `x` is the value of the Durbin-Watson statistic, `m` the number of
+elements in `a`, and `n` the number of approximation terms (see Farebrother, 1980, eq. 5).
+
+# References
+
+  * J. Durbin and G. S. Watson, 1971, "Testing for Serial Correlation in Least Squares
+  Regression: III", Biometrika, Vol. 58, No. 1, pp. 1-19,
+  [http://www.jstor.org/stable/2334313](http://www.jstor.org/stable/2334313).
+  * R. W. Farebrother, 1980, "Algorithm AS 153: Pan's Procedure for the Tail Probabilities
+  of the Durbin-Watson Statistic", Journal of the Royal Statistical Society, Series C
+  (Applied Statistics), Vol. 29, No. 2, pp. 224-227,
+  [http://www.jstor.org/stable/2986316](http://www.jstor.org/stable/2986316).
+
+"""
+function pan_algorithm(a::AbstractArray, x::Float64, m::Int, n::Int)
+
+    try # catch case where ν is empty
+        ν = find(a.>=x)[1]
+        if ν == 1
+            sum = 0.0
+        else
+            k = 1
+            ν = ν - 1
+            h = m - ν
+
+            if ν <= h
+                d  = 2; h  = ν; k  = - k; j1 = 0; j2 = 2; j3 = 3; j4 = 1
+            else
+                d  = - 2; ν  = ν + 2; j1 = m - 2; j2 = m - 1; j3 = m + 1; j4 = m
+            end
+
+            pin = pi / (2n)
+            sum = 0.5 * (k + 1)
```

Comment:
Typically it's preferred to avoid float literals in computations such as this, since float literals are always `Float64`, which will promote smaller float types to `Float64` when they're used together in a computation. So in this case it would be `sum = (k + 1) / 2` (though also note that it doesn't _really_ matter, since that will promote to `Float64` anyway, but the general point still stands 😛). It might be worthwhile to check `@code_warntype pan_algorithm(<some args>)` to see whether the function remains type-stable.